### PR TITLE
ddl: Added check for partition name ending with space

### LINKED ISSUE #31535: Incorrect partition name is not resulting error in Tidb.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
drop table if exists t1;
create table t1 (a int) partition by list(a) (partition `s1 s2 ` values in (0), partition `s1 s2` values in (2));
USE information_schema;
SELECT table_name,partition_name  FROM partitions WHERE table_schema='test' AND table_name='t1' and partition_name ='s1 s2';
use test;
```

### 2. What did you expect to see? (Required)
```
mysql> create table t1 (a int) partition by list(a) (partition `s1 s2 ` values in (0), partition `s1 s2` values in (2));
ERROR 1567 (HY000): Incorrect partition name

mysql> SELECT table_name,partition_name  FROM partitions WHERE table_schema='test' AND table_name='t1' and partition_name ='s1 s2';
Empty set (0.00 sec)
```
### 3. What did you see instead (Required)
```sql
mysql> create table t1 (a int) partition by list(a) (partition `s1 s2 ` values in (0), partition `s1 s2` values in (2));
Query OK, 0 rows affected (0.01 sec)

mysql> SELECT table_name,partition_name  FROM partitions WHERE table_schema='test' AND table_name='t1' and partition_name ='s1 s2';
+------------+----------------+
| table_name | partition_name |
+------------+----------------+
| t1         | s1 s2          |
| t1         | s1 s2          |
+------------+----------------+
2 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```sql
| Release Version: v5.4.0-alpha-264-g6efa36df6
Edition: Community
Git Commit Hash: 6efa36df6cff325106f67ecfe3d79816ba37ca6a
Git Branch: master
UTC Build Time: 2021-12-28 02:03:55
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #31535

Problem Summary:

Missed to check partition name as valid Identifier (like not ending with a space).

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
No longer allowing partition names ending with space.
```

### CODE REVIEW

**Review by djshow832 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -6935,3 +6935,20 @@ func (d *ddl) AlterTableNoCache(ctx sessionctx.Context, ti ast.Ident) (err error
 	err = d.doDDLJob(ctx, job)
 	return d.callHookOnChanged(err)
 }
+
+// IsValidIdentifier checks if the identifier is correct.
+// See https://dev.mysql.com/doc/refman/5.7/en/identifiers.html
+func IsValidIdentifier(name string) bool {
```
Comment (tiancaiamao) [inline ddl/ddl_api.go:6941]: Why should the ddl package expose an API named `IsValidIdentifier`?

**Review by tiancaiamao — COMMENTED** [review]
IMHO, it's better to tiny update the old `isIncorrectName()` function, and fix this check in the preprocess:

```
git diff
diff --git a/planner/core/preprocess.go b/planner/core/preprocess.go
index 9e4ddbf20..72a1698d0 100644
--- a/planner/core/preprocess.go
+++ b/planner/core/preprocess.go
@@ -837,6 +837,16 @@ func (p *preprocessor) checkCreateTableGrammar(stmt *ast.CreateTableStmt) {
                p.err = ddl.ErrTableMustHaveColumns
                return
        }
+
+       if stmt.Partition != nil {
+               for _, def := range stmt.Partition.Definitions {
+                       pName := def.Name.String()
+                       if isIncorrectName(pName) {
+                               p.err = ddl.ErrWrongPartitionName.GenWithStackByArgs(pName)
+                               return
+                       }
+               }
+       }
 }
```

**Comment by mjonss** [comment]
> IMHO, it's better to tiny update the old `isIncorrectName()` function, and fix this check in the preprocess:

Thanks for the suggestion, I agree and have updated the patch.
The reason for doing it the other way was to try to cover any code path creating a new partition, but currently your suggestion is better.

File: ddl/db_integration_test.go [hunk 2]
```diff
@@ -3821,3 +3821,24 @@ func TestIssue29326(t *testing.T) {
 	err = tk.ExecToErr("create definer=`root`@`127.0.0.1` view v1 as select t1.id, t2.id from t1,t2 where t1.id=t2.id")
 	require.True(t, infoschema.ErrColumnExists.Equal(err))
 }
+
+func (s *testIntegrationSuite3) TestInvalidPartitionNameWhenCreateTable(c *C) {
```
Comment (hawkingrei) [inline ddl/db_integration_test.go:3825]: @mjonss You need to use new test framework to implement it. those test cases have been migrated.

**Pull request merged**